File: ag-ui/apps/dojo/src/mastra/agents/shared-state.ts
import { Agent } from "@mastra/core/agent";
import { Memory } from "@mastra/memory";
import { z } from "zod";
import { getStorage } from "../storage";

export const sharedStateAgent = new Agent({
  id: "shared_state",
  name: "shared_state",
  instructions: `
    You are a helpful assistant for creating recipes.

    The recipe in working memory is the user's CURRENT state — they edit it directly
    in the UI. Treat it as the source of truth, even when an earlier message said
    otherwise.

    IMPORTANT:
    1. Create a recipe using the existing ingredients and instructions. Make sure the recipe is complete.
    2. For ingredients, append new ingredients to the existing ones.
    3. For instructions, append new steps to the existing ones.
    4. 'ingredients' is always an array of objects with 'icon', 'name', and 'amount' fields
    5. 'instructions' is always an array of strings
    6. 'special_preferences', 'skill_level' and 'cooking_time' are chosen by the USER,
       not you. NEVER add, remove, or change them — leave them exactly as they are in
       working memory. In particular, do NOT re-add a preference the user removed, and
       do NOT add "Spicy" (or any preference) on your own.
    7. Instead, make 'ingredients' and 'instructions' MATCH the user's
       'special_preferences': if "Vegan" is set, use vegan ingredients; if "Spicy" is
       set, add heat; and if a preference is NOT set, make sure the recipe does not
       reflect it (e.g. if "Spicy" is absent, use no chili/heat).

    If you have just created or modified the recipe, just answer in one sentence what you did. Do not describe the recipe, just say what you did. Do not mention "working memory", "memory", or "state" in your answer.
  `,
  model: "openai/gpt-4.1-mini",
  memory: new Memory({
    storage: getStorage(),
    options: {
      workingMemory: {
        enabled: true,
        schema: z.object({
          recipe: z.object({
            skill_level: z
              .enum(["Beginner", "Intermediate", "Advanced"])
              .describe("The skill level required for the recipe"),
            special_preferences: z
              .array(
                z.enum([
                  "High Protein",
                  "Low Carb",
                  "Spicy",
                  "Budget-Friendly",
                  "One-Pot Meal",
                  "Vegetarian",
                  "Vegan",
                ]),
              )
              .describe("A list of special preferences for the recipe"),
            cooking_time: z
              .enum(["5 min", "15 min", "30 min", "45 min", "60+ min"])
              .describe("The cooking time of the recipe"),
            ingredients: z
              .array(
                z.object({
                  icon: z
                    .string()
                    .describe(
                      "The icon emoji (not emoji code like '\\x1f35e', but the actual emoji like 🥕) of the ingredient",
                    ),
                  name: z.string().describe("The name of the ingredient"),
                  amount: z.string().describe("The amount of the ingredient"),
                }),
              )
              .describe(
                "Entire list of ingredients for the recipe, including the new ingredients and the ones that are already in the recipe",
              ),
            instructions: z
              .array(z.string())
              .describe(
                "Entire list of instructions for the recipe, including the new instructions and the ones that are already there",
              ),
          }),
        }),
      },
    },
  }),
});

File: ag-ui/integrations/ag2/python/examples/server/api/shared_state.py
"""Shared State feature.

Recipe assistant that maintains shared state (recipe) across turns.
Uses ContextVariables and ReplyResult like agentic_generative_ui.
"""

from enum import StrEnum
from textwrap import dedent

from autogen import ConversableAgent, LLMConfig
from autogen.ag_ui import AGUIStream
from autogen.agentchat import ContextVariables, ReplyResult
from autogen.tools import tool
from fastapi import FastAPI
from pydantic import BaseModel, Field


class SkillLevel(StrEnum):
    """The level of skill required for the recipe."""

    BEGINNER = "Beginner"
    INTERMEDIATE = "Intermediate"
    ADVANCED = "Advanced"


class SpecialPreferences(StrEnum):
    """Special preferences for the recipe."""

    HIGH_PROTEIN = "High Protein"
    LOW_CARB = "Low Carb"
    SPICY = "Spicy"
    BUDGET_FRIENDLY = "Budget-Friendly"
    ONE_POT_MEAL = "One-Pot Meal"
    VEGETARIAN = "Vegetarian"
    VEGAN = "Vegan"


class CookingTime(StrEnum):
    """The cooking time of the recipe."""

    FIVE_MIN = "5 min"
    FIFTEEN_MIN = "15 min"
    THIRTY_MIN = "30 min"
    FORTY_FIVE_MIN = "45 min"
    SIXTY_PLUS_MIN = "60+ min"


class Ingredient(BaseModel):
    """A class representing an ingredient in a recipe."""

    icon: str = Field(
        default="ingredient",
        description="The icon emoji (e.g. 🥕) of the ingredient",
    )
    name: str = Field(description="Name of the ingredient")
    amount: str = Field(description="Amount of the ingredient")


class Recipe(BaseModel):
    """A class representing a recipe."""

    skill_level: SkillLevel = Field(
        default=SkillLevel.BEGINNER,
        description="The skill level required for the recipe",
    )
    special_preferences: list[SpecialPreferences] = Field(
        default_factory=list,
        description="Any special preferences for the recipe",
    )
    cooking_time: CookingTime = Field(
        default=CookingTime.FIVE_MIN,
        description="The cooking time of the recipe",
    )
    ingredients: list[Ingredient] = Field(
        default_factory=list,
        description="Ingredients for the recipe",
    )
    instructions: list[str] = Field(
        default_factory=list,
        description="Instructions for the recipe",
    )


class RecipeSnapshot(BaseModel):
    """A class representing the state of the recipe."""

    recipe: Recipe = Field(
        default_factory=Recipe,
        description="The current state of the recipe",
    )


@tool()
async def get_current_recipe(context_variables: ContextVariables) -> str:
    """Return the current recipe state as JSON so you can read it before updating.

    Call this when you need to see the existing recipe (e.g. ingredients, instructions)
    before making changes or when the user asks to modify the recipe.
    """
    data = context_variables.data
    if not data:
        return RecipeSnapshot().model_dump_json(indent=2)
    snapshot = RecipeSnapshot.model_validate(data)
    return snapshot.model_dump_json(indent=2)


@tool()
async def display_recipe(
    context_variables: ContextVariables,
    recipe: Recipe,
) -> ReplyResult:
    """Display the recipe to the user.

    Use this to present the full recipe (or an updated version) to the user.
    Append new ingredients to existing ones when extending the recipe.
    Do not repeat the recipe in your message after calling this tool.

    Args:
        recipe: The recipe to display (full snapshot including ingredients and instructions).
    """
    snapshot = RecipeSnapshot(recipe=recipe)
    context_variables.update(snapshot.model_dump())
    return ReplyResult(
        message="Recipe displayed",
        context_variables=context_variables,
    )


agent = ConversableAgent(
    name="recipe_assistant",
    system_message=dedent("""
        You are a helpful assistant for creating recipes.

        IMPORTANT:
        - Create a complete recipe using the existing ingredients
        - Append new ingredients to the existing ones
        - Use the `display_recipe` tool to present the recipe to the user
        - Do NOT repeat the recipe in the message, use the tool instead
        - Do NOT run the `display_recipe` tool multiple times in a row

        Once you have created the updated recipe and displayed it to the user,
        summarise the changes in one sentence, don't describe the recipe in
        detail or send it as a message to the user.
    """),
    llm_config=LLMConfig({"model": "gpt-4o-mini", "stream": True}),
    human_input_mode="NEVER",
    functions=[get_current_recipe, display_recipe],
)

stream = AGUIStream(agent)
shared_state_app = FastAPI()
shared_state_app.mount("", stream.build_asgi())
